packages/tpetra/core/src/Tpetra_MultiVector_def.hpp

```diff
@@ -3295,34 +3296,64 @@ namespace Tpetra {
     const char prefix[] = "Tpetra::MultiVector constructor (offsetView): ";
+    const char suffix[] = "Please report this bug to the Tpetra developers.";
+    int lclGood = 1;
+    int gblGood = 1;    
+    std::unique_ptr<std::ostringstream> errStrm;
+    const bool debug = ::Tpetra::Details::Behavior::debug ();
+    const bool verbose = ::Tpetra::Details::Behavior::verbose ();
+
+    // Be careful to use the input Map's communicator, not X's.  The
+    // idea is that, on return, *this is a subview of X, using the
+    // input Map.
+    const auto comm = subMap.getComm ();
+    TEUCHOS_ASSERT( ! comm.is_null () );
+    const int myRank = comm->getRank ();
 
+    const size_t lclNumRowsBefore = X.getLocalLength ();
+    const size_t numCols = X.getNumVectors ();
     const size_t newNumRows = subMap.getNodeNumElements ();
+    if (verbose) {
+      std::ostringstream os;
+      os << "Proc " << myRank << ": " << prefix << "X: {lclNumRows: "
+	 << lclNumRowsBefore << ", origLclNumRows: " << X.getOrigNumLocalRows ()
+	 << ", numCols: " << numCols << "}, subMap: " << newNumRows << endl;
+      std::cerr << os.str ();
+    }
+    // We ask for the _original_ number of rows in X, because X could
+    // be a shorter (fewer rows) view of a longer MV.  (For example, X
+    // could be a domain Map view of a column Map MV.)
     const bool tooManyElts = newNumRows + offset > X.getOrigNumLocalRows ();
     if (tooManyElts) {
-      const int myRank = X.getMap ()->getComm ()->getRank ();
-      TEUCHOS_TEST_FOR_EXCEPTION(
-        newNumRows + offset > X.getLocalLength (), std::runtime_error,
-        prefix << "Invalid input Map.  The input Map owns " << newNumRows <<
-        " entries on process " << myRank << ".  offset = " << offset << ".  "
-        "Yet, the MultiVector contains only " << X.getOrigNumLocalRows () <<
-        " rows on this process.");
+      errStrm = std::unique_ptr<std::ostringstream> (new std::ostringstream);
+      *errStrm << "  Proc " << myRank << ": subMap.getNodeNumElements() (="
+	<< newNumRows << ") + offset (=" << offset << ") > X.getOrigNumLocal"
+        "Rows() (=" << X.getOrigNumLocalRows () << ")." << endl;
+      lclGood = 0;
+      TEUCHOS_TEST_FOR_EXCEPTION
+	(! debug && tooManyElts, std::invalid_argument,
+	 prefix << errStrm->str () << suffix);
     }
 
-    // mfh 27 Sep 2017: This debugging code depends on being able to
-    // declare variables that we can use below, so it's too hard for
-    // now to use run-time control to enable this code.
-#ifdef HAVE_TPETRA_DEBUG
-    const size_t strideBefore =
-      X.isConstantStride () ? X.getStride () : static_cast<size_t> (0);
-    const size_t lclNumRowsBefore = X.getLocalLength ();
-    const size_t numColsBefore = X.getNumVectors ();
-    const impl_scalar_type* hostPtrBefore =
-      X.getDualView().view_host().data ();
-#endif // HAVE_TPETRA_DEBUG
+    if (debug) {
+      reduceAll<int, int> (*comm, REDUCE_MIN, lclGood, outArg (gblGood));
```

Comment:
Since `blGood` is really an in/out arg, this might be better to use [`Teuchos::inoutArg()`](https://trilinos.org/docs/dev/packages/teuchos/doc/html/classTeuchos_1_1Ptr.html#acb9418d4e46ac7b589f8de2ab89bc56e) to make this more self-documenting.  Just a suggestion.

Reply:
@bartlettroscoe Actually, it's totally an output argument; its value on input doesn't matter.  `lclGood` is the input value here; their minimum over all processes in `*comm` is written to `gblGood`.

Reply:
@mhoemmen, ah, excellent!  It has been a long time since I looked at `reduceAll()`.  I thought that it took the initial value and reduced them all and put the result back in that object.  (Am I thinking about `broadcast()`?)  Now I see the `iclGood` argument.  Thanks!